### ISSUE TITLE
Dirty write prevention doesn't protect an external change that leaves filesize unaltered

### ISSUE THREAD

**Issue by gjsjohnmurray** [microsoft/vscode#119002]
I'm trying to understand this code and its extensive comment:

https://github.com/microsoft/vscode/blob/83b6cdc610703902055aa42984869ff46227126b/src/vs/platform/files/common/fileService.ts#L400-L418

Try the following:
1. Open an existing text file in VS Code.
2. Make an obvious change to it (e.g. insert a line of additional text)  but do not save the dirty file.
3. Open the same file in another editor (I used Notepad).
4. In that editor make a change that doesn't alter the file's length (e.g. replace a lowercase letter with an uppercase one). Save this change.
5. Back in VS Code, save the change you made in step 2.
6. :bug: No dirty write warning appears. Instead the change made in step 4 is lost.

Version: 1.55.0-insider (user setup)
Commit: 84fe402d655e029eb1a5c04e675bf64788fa7fcf
Date: 2021-03-15T05:14:29.836Z
Electron: 11.3.0
Chrome: 87.0.4280.141
Node.js: 12.18.3
V8: 8.7.220.31-electron.0
OS: Windows_NT x64 10.0.19042

**Comment by bpasero** [comment]
👏  for finding this, but I am inclined to close this was wont fix because there is reasons to also check on the content size and not just modification time. I had seen a lot of issues where modification time was advanced but the content not changed (aka false positives) so I ended up with this solution. 

We could maybe open up https://github.com/microsoft/vscode/issues/919 to cover this case as well. If we had a hash over the contents of the file we could use it to compare the content properly and not just through the size.

**Comment by gjsjohnmurray** [comment]
For dirty write protection I think false positives are preferable to false negatives.

Can we compute and store a content hash when reading the file, then if the file to be overwritten has a later timestamp but the same size we'd fetch it and compute the hash, triggering dirty write protection if the hash has changed?

**Comment by bpasero** [comment]
> For dirty write protection I think false positives are preferable to false negatives

I remember in the early days this was very frustrating for users because showing this notification prevents you from normal save and it is hard to understand why that happens. This was especially happening with remote network drives where maybe the file system driver was causing issues with the `mtime` on disk and VSCode was constantly thinking the file has been modified. Or even worse, a file watcher or some other external program bumping the `mtime` without changing the content.

> Can we compute and store a content hash when reading the file, then if the file to be overwritten has a later timestamp but the same size we'd fetch it and compute the hash, triggering dirty write protection if the hash has changed?

Yes that is what I had thought https://github.com/microsoft/vscode/issues/919 could solve but we can leave this as individual feature request. Bottom line: it can be expensive to compute this hash every time you open a file so preferably I would only compute the hash when we run into the issue of `mtime` having advanced while the size is the same.

**Comment by gjsjohnmurray** [comment]
> preferably I would only compute the hash when we run into the issue of `mtime` having advanced while the size is the same.

Won't the hash of the originally-opened file need to be computed at load time? Or does VS Code keep a local copy of how this looked before the user made any changes?

Another suggestion is to make this behaviour configurable, perhaps by scheme or scheme+authority. Then maybe make the code default to etag-only comparison when etags are being supplied, else the combination of mtime and size. People who start getting false positives would then add a setting to revert to the previous behaviour for the affected scheme or scheme+authority.

**Comment by bpasero** [comment]
@gjsjohnmurray yeah good point, it would need to be a hash of the contents on disk. However, we do have a undo/redo stack so we could possibly figure out the original contents all the way back by applying the undo steps on the model if we wanted to compute this lazily. 

I think I would rather want a solution like the hash one for any file system provider and rather not have this configurable.

**Comment by joaomoreno** [comment]
I wish we had a more sensible default here:

- For most cases mtime is enough, this should be the default case
- There should be a behavior state per FS which would enable additional content checking, whenever mtime gets invalidated
- There should be a setting to manually enable that behavior, at user/workspace/folder level

**Comment by bpasero** [comment]
We have a setting for the user to control our dirty write prevention called `files.saveConflictResolution`:

![image](https://user-images.githubusercontent.com/900690/186865362-ea9cdb31-c7e7-4b31-a565-662724430791.png)

I think my statement from https://github.com/microsoft/vscode/issues/119002#issuecomment-799992944 is still correct, only do the more expensive content hash when `size` did not change but `mtime` advanced.

**Comment by gjsjohnmurray** [comment]
Could we use a heuristic to detect when that appears to be happening, then stop doing the check for that folder"s files?
